AutoGraph/WebSocketClient.swift

```diff
@@ -0,0 +1,300 @@
+        guard self.state == .connected else {
+            throw WebSocketError.webSocketNotConnected(subscriptionPayload: subscriptionPayload)
+        }
+        self.write(subscriptionPayload)
+    }
+    
+    private var reconnecting = false
+    /// Attempts to reconnect and re-subscribe with multiplied backoff up to 30 seconds. Returns the delay.
+    func reconnect() -> DispatchTimeInterval? {
+        guard !self.reconnecting, !self.fullDisconnect else { return nil }
+        if self.attemptReconnectCount > 0 {
+            self.attemptReconnectCount -= 1
+        }
+        self.reconnecting = true
+        
+        let delayInSeconds = DispatchTimeInterval.seconds(min(abs(kAttemptReconnectCount - self.attemptReconnectCount) * 10, 30))
+        DispatchQueue.main.asyncAfter(deadline: .now() + delayInSeconds) { [weak self] in
+            guard let self = self else { return }
+            // Requeue all so they don't get error callbacks on disconnect and they get re-subscribed on connect.
+            self.requeueAllSubscribers()
+            self.disconnectAndPossiblyReconnect()
+            self.webSocket.connect()
+        }
+        return delayInSeconds
+    }
+    
+    /// Takes all subscriptions and puts them back on the queue, used for reconnection.
+    func requeueAllSubscribers() {
+        for (_, subscriptionSet) in self.subscriptions {
+            for (subscriber, subscriptionResponseHandler) in subscriptionSet {
+                self.queuedSubscriptions[subscriber] = connectionCompletionBlock(subscriber: subscriber, responseHandler: subscriptionResponseHandler)
+            }
+        }
+        self.subscriptions.removeAll(keepingCapacity: true)
+    }
+    
+    // TODO: test
+    /// Take all connection completion blocks out of the queue and runs them. Removes from queue first to avoid any side affects.
+    func didConnect() throws {
+        self.state = .connected
+        self.attemptReconnectCount = kAttemptReconnectCount
+        self.reconnecting = false
+        
+        let connectedPayload = try GraphQLWSProtocol.connectionInit.serializedSubscriptionPayload()
+        self.write(connectedPayload)
+        
+        let queuedSubscriptions = self.queuedSubscriptions
+        self.queuedSubscriptions.removeAll()
+        queuedSubscriptions.forEach { (_, connected: WebSocketConnected) in
+            connected(.success(()))
+        }
+    }
+    
+    func reset() {
+        self.disconnect()
+        self.subscriptions.removeAll()
+        self.queuedSubscriptions.removeAll()
+        self.attemptReconnectCount = kAttemptReconnectCount
+        self.reconnecting = false
+    }
+}
+
+// MARK: - Class Method
+
+extension WebSocketClient {
+    class func connectionRequest(url: URL) throws -> URLRequest {
+        var defaultHeders = [String: String]()
+        defaultHeders["Sec-WebSocket-Protocol"] = "graphql-ws"
+        defaultHeders["Origin"] = url.absoluteString
```

Comment:
Okay so problem with this. If we generate outside the module and pass it in then the origin could potentially be wrong so then the connection will fail. The origin has to be the base url with no path extensions. So we we pass in a url with `wss://api.remind.com/subscriptions` then the origin would not be `https://api.remind.com` 

So what we can do is something like this 

`defaultHeader["Origin"] = url.deletingLastPathComponent().absoluteString.replacingOccurrences(of: "wss", with: "https")`

Reply:
@rex-remind101

Reply:
But that is always expecting on path component and the protocol is wss

Reply:
we should have the user pass in headers if they want to. we should have `setHeader` be a function on the `WebSocketClient`

Reply:
Then we can remove this origin and just use the user's headers